src/Orleans/AssemblyLoader/AssemblyLoader.cs

```diff
@@ -104,33 +104,57 @@ internal class AssemblyLoader
 
         public static T LoadAndCreateInstance<T>(string assemblyName, TraceLogger logger) where T : class
         {
-            var exeRoot = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
-            var dirs = new Dictionary<string, SearchOption>
+            bool alreadyLoggedError = false;
+            try
             {
-                {exeRoot, SearchOption.TopDirectoryOnly}
-            };
+                var exeRoot = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+                var dirs = new Dictionary<string, SearchOption>
+                {
+                    {exeRoot, SearchOption.TopDirectoryOnly}
+                };
 
-            AssemblyLoaderPathNameCriterion[] includeCriteria =
-            {
-                AssemblyLoaderCriteria.IncludeFileNames(new[] {assemblyName})
-            };
-            AssemblyLoaderReflectionCriterion[] loadCriteria =
-            {
-                AssemblyLoaderCriteria.LoadTypesAssignableFrom(typeof (T))
-            };
+                AssemblyLoaderPathNameCriterion[] includeCriteria =
+                {
+                    AssemblyLoaderCriteria.IncludeFileNames(new[] {assemblyName})
+                };
+                AssemblyLoaderReflectionCriterion[] loadCriteria =
+                {
+                    AssemblyLoaderCriteria.LoadTypesAssignableFrom(typeof (T))
+                };
 
-            var discoveredAssemblyLocations = LoadAssemblies(dirs, includeCriteria, loadCriteria, logger);
+                var discoveredAssemblyLocations = LoadAssemblies(dirs, includeCriteria, loadCriteria, logger);
 
-            if (discoveredAssemblyLocations.Count == 0)
-                throw new TypeLoadException("Type " + typeof(T).Name + " wasn't found");
+                if (discoveredAssemblyLocations.Count == 0)
+                {
+                    var error = String.Format("Failed LoadAndCreateInstance for type {0} from assembly {1}. " +
+                                              "Make sure this assembly is in your current executing directory and the type is defined there.", 
+                        typeof(T).FullName, assemblyName);
+                    logger.Error(ErrorCode.Loader_LoadAndCreateInstance_Failure_1, error);
+                    alreadyLoggedError = true;
+                    throw new TypeLoadException(error);
+                }
 
-            if (discoveredAssemblyLocations.Count > 1)
-                throw new TypeLoadException("Type " + typeof(T).Name + " was found more than once");
+                if (discoveredAssemblyLocations.Count > 1)
+                {
+                    var error = String.Format("Type {0} was found more than once in assembly {1}.", typeof(T).FullName, assemblyName);
+                    logger.Error(ErrorCode.Loader_LoadAndCreateInstance_Failure_2, error);
+                    alreadyLoggedError = true;
+                    throw new TypeLoadException(error);
```

Comment:
It's better to throw an OrleansException instead of TypeLoadException here, so that it's more clear that it didn't come from CLR.